sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -184,6 +202,23 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
     case a @ Aggregate(_, _, p @ Project(_, Join(_, right, RightOuter, _, _)))
         if a.groupOnly && p.references.subsetOf(right.outputSet) =>
       a.copy(child = p.copy(child = right))
+
+    case a @ Aggregate(_, aggExprs, Join(left, _, LeftOuter, _, _))
+        if a.references.subsetOf(left.outputSet) &&
+          allDuplicateAgnostic(aggExprs) =>
+      a.copy(child = left)
+    case a @ Aggregate(_, aggExprs, Join(_, right, RightOuter, _, _))
+        if a.references.subsetOf(right.outputSet) &&
+          allDuplicateAgnostic(aggExprs) =>
+      a.copy(child = right)
+    case a @ Aggregate(_, aggExprs, p @ Project(_, Join(left, _, LeftOuter, _, _)))
```

Comment:
This looks a bit risky. What if the project contains expensive or nondeterministic expressions?

Reply:
We would pull out nondeterministic to a new project, so some thing like this ? 
```sql
SELECT rand(), max(c2) FROM (
  SELECT t1.c1, t1.c2 FROM t1 LEFT JOIN t2 ON t1.c1 = t2.c1
)
GROUP BY rand()
```

How about adding a condition `a.groupingExpressions.forall(_.deterministic)` ?

Reply:
Change to `projectList.forall(_.deterministic)`, and add the test
```scala
val p1 = x.join(y, joinType, Some($"x.a" === $"y.a")).select($"$t.a" as "a1", rand(1) as "b1")
        .groupBy($"b1")($"b1", max($"a1")).analyze
comparePlans(Optimize.execute(p1), p1)
```